### PR TITLE
[IMP] base,account: manage subsidiary companies

### PR DESCRIPTION
**Pull request by fpodoo**
Allow sharing records between company
* accounts
* taxes
* fiscal positions

These records can be read and used in children companies.

This can be used to
* have different branding for different businesses
* allow more complex security rules
* consolidate branches differently
* manage different tax reports with different tax ids in the same
  country

[task-3371677](https://www.odoo.com/web#id=3371677&model=project.task)

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -3659,23 +3660,24 @@ def _check_company(self, fnames=None):
...
                     property_fields.append(name)
+            elif field.name == 'company_id':
+                company_changed = True
 
-        if not (regular_fields or property_fields):
+        if not (regular_fields or property_fields or company_changed):
             return
 
         inconsistencies = []
+        IrRule = self.env['ir.rule'].sudo()
```
Comment (qdp-odoo) [inline odoo/models.py:3670]: why the sudo()?

File: odoo/models.py [hunk 2]
```diff
@@ -3694,9 +3696,12 @@ def _check_company(self, fnames=None):
             lines = [_("Incompatible companies on records:")]
             company_msg = _lt("- Record is company %(company)r and %(field)r (%(fname)s: %(values)s) belongs to another company.")
             record_msg = _lt("- %(record)r belongs to company %(company)r and %(field)r (%(fname)s: %(values)s) belongs to another company.")
+            root_company_msg = _lt("- Only a root company can be set on %(record)r. Currently set to %(company)r")
             for record, name, corecords in inconsistencies[:5]:
                 if record._name == 'res.company':
                     msg, company = company_msg, record
+                elif record == corecords and name == 'company_id':
+                    msg, company = root_company_msg, record.company_id
```
Comment (qdp-odoo) [inline odoo/models.py:3704]: only 1 level of hierarchy allowed? :suspect:

File: odoo/models.py [hunk 3]
```diff
@@ -4696,6 +4701,10 @@ def _where_calc(self, domain, active_test=True):
         :return: the query expressing the given domain as provided in domain
         :rtype: Query
         """
+        if self.env.context.get('company_test'):
```
Comment (qdp-odoo) [inline odoo/models.py:4704]: doc

File: odoo/models.py [hunk 4]
```diff
@@ -5481,6 +5490,14 @@ def with_company(self, company):
 
         return self.with_context(allowed_company_ids=allowed_company_ids)
 
+    def with_only_company(self, company):
+        if company and not int(company):
+            return self
```
Comment (qdp-odoo) [inline odoo/models.py:5495]: gnii?

File: odoo/models.py [hunk 5]
```diff
@@ -5481,6 +5490,14 @@ def with_company(self, company):
 
         return self.with_context(allowed_company_ids=allowed_company_ids)
 
+    def with_only_company(self, company):
```
Comment (qdp-odoo) [inline odoo/models.py:5493]: doc

File: odoo/models.py [hunk 6]
```diff
@@ -5481,6 +5490,14 @@ def with_company(self, company):
...
+        if company and not int(company):
+            return self
+        allowed_company_ids = [int(company)]
+        if allowed_company_ids == self.env.context.get('allowed_company_ids') and self.env.context.get('company_test'):
+            return self
+        return self.with_context(allowed_company_ids=allowed_company_ids, company_test=True)
```
Comment (qdp-odoo) [inline odoo/models.py:5499]: brain has melted, need a new one and re-read that... :exploding_head:

File: odoo/models.py [hunk 7]
```diff
@@ -5712,7 +5729,14 @@ def filtered_domain(self, domain):
...
                 if comparator in ('child_of', 'parent_of'):
-                    stack.append(set(self.search([('id', 'in', self.ids), leaf], order='id')._ids))
+                    if key == 'company_id':  # avoid an explicit search
+                        value_companies = self.env['res.company'].browse(value)
+                        if comparator == 'child_of':
+                            stack.append({record.id for record in self if record.company_id.parent_ids & value_companies})
+                        else:
+                            stack.append({record.id for record in self if record.company_id & value_companies.parent_ids})
+                    else:
+                        stack.append(set(self.with_context(active_test=False).search([('id', 'in', self.ids), leaf], order='id')._ids))
```
Comment (qdp-odoo) [inline odoo/models.py:5739]: why the active_test=False ?

File: odoo/osv/expression.py [hunk 8]
```diff
@@ -896,11 +897,12 @@ def parent_of_domain(left, ids, left_model, parent=None, prefix=''):
                 (when available), or as an expanded [(left,in,parent_ids)] """
             if not ids:
                 return [FALSE_LEAF]
+            left_model = left_model.with_context(active_test=False)
```
Comment (qdp-odoo) [inline odoo/osv/expression.py:900]: why the active_test=False?

File: odoo/osv/expression.py [hunk 9]
```diff
@@ -896,11 +897,12 @@ def parent_of_domain(left, ids, left_model, parent=None, prefix=''):
...
             if left_model._parent_store:
                 parent_ids = [
                     int(label)
                     for rec in left_model.sudo().browse(ids)
-                    for label in rec.parent_path.split('/')[:-1]
+                    for label in (rec.parent_path or "").split('/')[:-1]
```
Comment (qdp-odoo) [inline odoo/osv/expression.py:905]: looks like a bugfix that should be in its dedicated commit

File: odoo/tools/misc.py [hunk 10]
```diff
@@ -1012,7 +1012,7 @@ def clean_context(context):
     """ This function take a dictionary and remove each entry with its key
     starting with ``default_``
     """
-    return {k: v for k, v in context.items() if not k.startswith('default_')}
+    return {k: v for k, v in context.items() if not k.startswith('default_') and not k == 'company_test'}
```
Comment (qdp-odoo) [inline odoo/tools/misc.py:1015]: mmh not super clean to hack in there for testing purposes. It would be better to extend this method in the tests

**Review by qdp-odoo — CHANGES_REQUESTED** [review]

File: odoo/models.py (hunk 1 above)
Comment (william-andre) [reply odoo/models.py:3670]: Because you need it to be able to access the rules

File: odoo/addons/test_http/tests/test_static.py [hunk 11]
```diff
@@ -265,10 +265,10 @@ def test_static15_range(self):
         )
 
     def test_static16_public_access_rights(self):
-        public_user = self.env.ref('base.public_user')
+        default_user = self.env.ref('base.default_user')
 
         with self.subTest('model access rights'):
-            res = self.url_open(f'/web/content/res.users/{public_user.id}/image_128')
+            res = self.url_open(f'/web/content/res.users/{default_user.id}/image_128')
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/test_http/tests/test_static.py:271]: Yes (along with the `active_test` stuff).
Asked the ORM team many times about it, only answer I got was "do what you think is right, you probably know what you're doing"

File: odoo/addons/base/security/base_security.xml [hunk 12]
```diff
@@ -152,6 +155,14 @@
             <field name="domain_force">['|', ('share', '=', False), ('company_ids', 'in', company_ids)]</field>
         </record>
 
+        <record id="res_users_company_consistency" model="ir.rule">
+            <field name="name">user company_consistency</field>
+            <field name="model_id" ref="model_res_users"/>
+            <field name="domain_force">[('company_ids', 'in', company_ids)]</field>
+            <field name="company_consistency" eval="1"/>
+            <field name="active" eval="0"/>
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/security/base_security.xml:163]: No, because otherwise you wouldn't be able to see all internal users like the previous rule suggests.
This one basically only manages the `_check_company` (and the domain in the UI) between related records; but it won't prevent seeing the internal user generally. (you can still talk to them etc.)

File: odoo/addons/base/models/res_company.py [hunk 13]
```diff
@@ -290,3 +345,19 @@ def _get_main_company(self):
...
+        accessible = self.env.companies
+        return (children & accessible).ids
+
+    def _accessible_branches(self):
+        return self.browse(self.__accessible_branches())
+
+    def _all_branches_selected(self):
+        return (
+            len(self.env.companies.root_id) == 1
+            and self.env.companies == self.env['res.company'].sudo().search([('id', 'child_of', self.env.company.root_id.ids)])
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/res_company.py:362]: Because this check is general and doesn't depend on your rights regarding the main company/branches

It is used to disable some actions that only make sense if you have access (and access is activated) to all branches of a company

File: odoo/addons/base/models/res_company.py [hunk 14]
```diff
@@ -290,3 +345,19 @@ def _get_main_company(self):
             main_company = self.env['res.company'].sudo().search([], limit=1, order="id")
 
         return main_company
+
+    @ormcache('tuple(self.env.companies.ids)', 'self.id')
+    def __accessible_branches(self):
+        self.ensure_one()
+        children = self.sudo().search([('id', 'child_of', self.id)])
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/res_company.py:352]: Because this can also select a company that is not currently in your company selector, as long as you can have access to it

File: odoo/addons/base/models/res_company.py [hunk 15]
```diff
@@ -261,10 +312,14 @@ def _check_active(self):
                         active_users=company_active_users,
                     ))
 
-    @api.constrains('parent_id')
-    def _check_parent_id(self):
-        if not self._check_recursion():
-            raise ValidationError(_('You cannot create recursive companies.'))
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/res_company.py:267]: It is covered by `_parent_store`

File: odoo/addons/base/models/res_company.py [hunk 16]
```diff
@@ -91,10 +90,19 @@ def _get_company_address_field_names(self):
...
+
     def _get_company_address_update(self, partner):
         return dict((fname, partner[fname])
                     for fname in self._get_company_address_field_names())
-    
+
+    @api.depends('parent_path')
+    def _compute_parent_ids(self):
+        for company in self:
+            company.parent_ids = self.browse(int(id) for id in company.parent_path.split('/') if id) if company.parent_path else company
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/res_company.py:103]: No, I'm assigning a recordset; not a list of recordsets (or even worse: a list of ids, as it would possibly conflict with ORM commands)

File: odoo/addons/base/models/ir_rule.py [hunk 17]
```diff
@@ -119,6 +126,20 @@ def _get_rules(self, model_name, mode='read'):
         self._cr.execute(query, (model_name, self._uid))
         return self.browse(row[0] for row in self._cr.fetchall())
 
+    @tools.ormcache('model_name', 'placeholder', 'self.env.company.id')
+    def _compute_company_consistency_domain(self, model_name, placeholder=None):
+        model = self.env[model_name]
+        if 'company_id' in model._fields and (model._fields['company_id'].store or model._fields['company_id'].related):
+            domain = self.sudo().with_context(active_test=False).search([
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/ir_rule.py:133]: > if a record rule was disabled you don't want to find it back

Look at the rule for `res.users` for instance. The general rule is not always just depending on `company_id`, so we might need to have a fake (inactive) rule only for keeping the consistency between relational records

File: odoo/addons/base/models/ir_rule.py [hunk 18]
```diff
@@ -119,6 +126,20 @@ def _get_rules(self, model_name, mode='read'):
         self._cr.execute(query, (model_name, self._uid))
         return self.browse(row[0] for row in self._cr.fetchall())
 
+    @tools.ormcache('model_name', 'placeholder', 'self.env.company.id')
+    def _compute_company_consistency_domain(self, model_name, placeholder=None):
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/ir_rule.py:130]: Mostly to be able to send the domain to the UI, with something like `('company', '=', company_id)` instead of `('company', '=', 1)`

File: odoo/addons/base/models/res_company.py [hunk 19]
```diff
@@ -140,6 +148,15 @@ def _compute_uses_default_logo(self):
         for company in self:
             company.uses_default_logo = not company.logo or company.logo == default_logo
 
+    @api.depends('root_id')
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/addons/base/models/res_company.py:151]: It's not stored, and I'd like to avoid the trigger tree to have to traverse `res.partner` for something not really important

**Review by qdp-odoo — COMMENTED** [review]
very quick look

**Comment by qdp-odoo** [comment]
@robodoo rebase-ff r+

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

File: odoo/models.py [hunk 20]
```diff
@@ -3632,6 +3632,21 @@ def get_base_url(self):
...
+        elif isinstance(companies, BaseModel):
+            return companies.ids
+        elif isinstance(companies, (list, tuple)):
+            return companies
+        return [companies]
+
+    def _check_company_domain_parent_of(self, companies=None):
+        return ['|', ('company_id', '=', False), ('company_id', 'parent_of', self._parse_companies(companies))]
+
+    def check_company_domain(self, companies=None):
```
Comment (rco-odoo) [inline odoo/models.py:3648]: Please make the method private, as making it public has no purpose.
And add a docstring to document what it should be used for.
Also, I recommend to make the argument mandatory.  The parameter should be explicit IMHO.

File: odoo/models.py [hunk 21]
```diff
@@ -3671,12 +3686,10 @@ def _check_company(self, fnames=None):
             # with the company of the origin document, i.e. `self.account_id.company_id == self.company_id`
             for name in regular_fields:
                 corecord = record.sudo()[name]
-                # Special case with `res.users` since an user can belong to multiple companies.
-                if corecord._name == 'res.users' and corecord.company_ids:
-                    if not (company <= corecord.company_ids):
+                if corecord and ('company_id' in corecord._fields or corecord._name == 'res.company'):
```
Comment (rco-odoo) [inline odoo/models.py:3689]: I don't understand the second part of the `and`. The condition `'company_id' in corecord._fields` is implied by being in `regular_fields`, and the second one does not make sense to me.

File: odoo/fields.py [hunk 22]
```diff
@@ -2863,23 +2864,27 @@ def _related_domain(self):
...
                 # company_id comes from the id of the current record
                 cid = "id" if self.model_name == "res.company" else "company_id"
                 if self.comodel_name == "res.users":
                     # User allowed company ids = user.company_ids
-                    return f"['|', (not {cid}, '=', True), ('company_ids', 'in', [{cid}])]"
+                    company_domain = f"['|', (not {cid}, '=', True), ('company_ids', 'in', [{cid}])]"
                 else:
-                    return f"[('company_id', 'in', [{cid}, False])]"
-        return self.domain(env[self.model_name]) if callable(self.domain) else self.domain
+                    company_domain = env[self.comodel_name].check_company_domain(companies=unquote(f"{cid} or current_company_id"))
```
Comment (rco-odoo) [inline odoo/fields.py:2883]: If the company field in the form is not set, the domain should not constrain the value.
And dropping `current_company_id` here does not look correct to me.
The company field in the form should be the reference company.
The "current company" is only useful for getting a default value.

File: odoo/models.py (hunk 20 above)
Comment (william-andre) [reply odoo/models.py:3648]: It could be needed by the web client to build a domain correctly if needed

File: odoo/models.py (hunk 21 above)
Comment (william-andre) [reply odoo/models.py:3689]: It is because `check_company_domain` always returns a domain containing `company_id`; otherwise I have to manage the existence of the field in that function making it harder to override.

File: odoo/models.py (hunk 20 above)
Comment (rco-odoo) [reply odoo/models.py:3648]: A public method requires a different API (arguments cannot be records).
If you don't have a concrete case, it does not need to be public.

File: odoo/models.py [hunk 23]
```diff
@@ -3633,6 +3633,26 @@ def get_base_url(self):
...
+    def _parse_companies(self, companies):
+        if companies is None:
+            return self.env.company.ids
+        elif isinstance(companies, BaseModel):
+            return companies.ids
+        elif isinstance(companies, (list, tuple)):
+            return companies
+        return [companies]
+
+    def _check_company_domain_parent_of(self, companies=None):
```
Comment (rco-odoo) [inline odoo/models.py:3645]: Could you move this method outside of the class?
That's a component which is reused in many classes, but does not belong to `BaseModel`...

Same remark for `_parse_companies`: it's a helper, not a method to be overridden.
By the way, its name is confusing.  Its name should reflect _what_ it does, not _how_ it does it.
It converts `companies` into a list of company ids, and does not even "parse" anything.
I suggest `to_company_ids(...)` instead.

File: odoo/models.py (hunk 20 above)
Comment (william-andre) [reply odoo/models.py:3648]: The argument can be anything, including a list of ids

File: odoo/models.py [hunk 24]
```diff
@@ -3633,6 +3633,26 @@ def get_base_url(self):
             raise ValueError("Expected singleton or no record: %s" % self)
         return self.env['ir.config_parameter'].sudo().get_param('web.base.url')
 
+    def _parse_companies(self, companies):
+        if companies is None:
+            return self.env.company.ids
```
Comment (rco-odoo) [inline odoo/models.py:3638]: I strongly advocate against this kind of fallback, as it often hides programming errors.
The argument should be explicit IMHO.

File: odoo/models.py [hunk 25]
```diff
@@ -3633,6 +3633,26 @@ def get_base_url(self):
...
+        return [companies]
+
+    def _check_company_domain_parent_of(self, companies=None):
+        return ['|', ('company_id', '=', False), ('company_id', 'parent_of', self._parse_companies(companies))]
+
+    def check_company_domain(self, companies=None):
+        """Domain to be used for company consistency between records regarding this model.
+
+        :param companies: the allowed companies on the related record
+        :type companies: BaseModel or list or tuple or int
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/models.py]: Added it in the docstring

File: odoo/models.py [hunk 26]
```diff
@@ -3671,12 +3691,10 @@ def _check_company(self, fnames=None):
             # with the company of the origin document, i.e. `self.account_id.company_id == self.company_id`
             for name in regular_fields:
                 corecord = record.sudo()[name]
-                # Special case with `res.users` since an user can belong to multiple companies.
-                if corecord._name == 'res.users' and corecord.company_ids:
-                    if not (company <= corecord.company_ids):
+                if corecord:
+                    domain = corecord.check_company_domain(company)
+                    if company and domain and not corecord.filtered_domain(domain):
                         inconsistencies.append((record, name, corecord))
```
[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/models.py]: I don't remember why I put it like that. If the CI doesn't fail, good by me :+1:

[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/models.py]: I seriously doubt it was the case since this test is failing and is explicitly setting a record linked to a company on a record without a company (`user_id` belongs to `company_main`, explicitly set in the setup)
https://github.com/odoo/odoo/blob/ac9162727cf282227ea5f6c5b8be9baaea9bc287/addons/sales_team/tests/common.py#L94-L99

[the remark this answers is not in the document]
Comment (william-andre) [reply odoo/models.py]: You're right, I changed the `_check_company_domain` on `res.users`

**Review by rco-odoo — APPROVED** [review]
The changes in the ORM itself are fine to me.
The fact that the code dealing with users was removed and simply delegated to the `res.user` method `_check_company_domain` is really nice :+1:

**Comment by qdp-odoo** [comment]
@robodoo pre-shot r+ rebase-ff :cowboy_hat_face:

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by jjscarafia** [comment]
Thanks @fpodoo !
Love this improvement to be happening!

**Comment by jjscarafia** [comment]
@fpodoo @william-andre 
Great improvement!
Any plans to follow similar path with follow ups? (if I'm not wrong they are not "branches aware")
Thanks!

**Comment by william-andre** [comment]
From what I saw in the code, it should work (as long as you have the company and branches selected in the company selector). If it doesn't work, you are welcome to open a ticket.

However, I'm not sure what to expect from it. Customers might be confused about buying something from one company but getting the followup done by another company.
The branding/style/header/footer of the branches might also be completely different...

**Comment by jjscarafia** [comment]
hi @william-andre , thanks for the quick response.

You're right. Maybe I've tested it wrong or tested the wrong version.
It works, different than 'credit' field does (you only see all the debt If you're on the parent company with child companies selected (but not if you're on a child/branch company).
Credit/debit fields computes the debt for all the hierarchy (of companies that you can see, selected on the company switcher)

So, forget about this, we will try this approach.

Thanks!

Some images to show it:
* if on parent company, I see invoices of both branches
image

* on child company, I see invoices only of actual branch
image

image

File: addons/account/wizard/account_automatic_entry_wizard.py [hunk 27]
```diff
@@ -139,9 +141,9 @@ def default_get(self, fields):
             raise UserError(_('You can only change the period/account for posted journal items.'))
         if any(move_line.reconciled for move_line in move_line_ids):
             raise UserError(_('You can only change the period/account for items that are not yet reconciled.'))
-        if any(line.company_id != move_line_ids[0].company_id for line in move_line_ids):
+        if any(line.company_id.root_id != move_line_ids[0].company_id.root_id for line in move_line_ids):
             raise UserError(_('You cannot use this wizard on journal entries belonging to different companies.'))
-        res['company_id'] = move_line_ids[0].company_id.id
+        res['company_id'] = move_line_ids[0].company_id.root_id.id
```
Comment (ngochung207) [inline addons/account/wizard/account_automatic_entry_wizard.py:146]: Hi @william-andre 

Could you please explain this line of code to me?
Currently, it is not possible to record the branch's transactions in the branch's journal (the branch's journal cannot be found in the view because the company has been changed to the parent company).

Comment (william-andre) [reply addons/account/wizard/account_automatic_entry_wizard.py:146]: It is probably a mistake.
The goal was to able to manage automatic entries for multiple entries coming from different branches.
